pyteal/compiler/optimizer/optimizer.py

```diff
@@ -0,0 +1,89 @@
+        skip_ids (optional): the slot ids that should be skipped during
+            optimization. At the moment this includes:
+                1. reserved slots because they may have dependencies outside
+                the current application. For example, the 'gloads' opcode can
+                access the slots of other applications in the tx group.
+                2. global slots because they're outside the scope of global
+                optimizations, which only apply to the control flow graph of
+                a single subroutine.
+                3. slots used with dynamic scratch vars. These slots use
+                indirection by means of the 'stores' opcode and dependencies
+                can only be determined at runtime."""
+
+    def __init__(self, *, scratch_slots: bool = False, skip_ids: Set[int] = None):
+        self.scratch_slots = scratch_slots
+        self.skip_ids: Set[int] = skip_ids if skip_ids is not None else set()
+
+
+def _remove_extraneous_slot_access(start: TealBlock, remove: Set[int]):
+    def keep_op(op: TealOp):
+        if type(op) != TealOp or (op.op != Op.store and op.op != Op.load):
+            return True
+
+        return cast(int, op.args[0]) not in remove
+
+    for block in TealBlock.Iterate(start):
+        block.ops = list(filter(keep_op, block.ops))
+
+
+def _has_load_dependencies(cur_block: TealBlock, start: TealBlock, slot: int, pos: int):
+    for block in TealBlock.Iterate(start):
+        for i, op in enumerate(block.ops):
+            if block == cur_block and i == pos:
+                continue
+
+            if type(op) == TealOp and op.op == Op.load and op.args[0] == slot:
+                return True
+
+    return False
+
+
+def _apply_slot_to_stack(cur_block: TealBlock, start: TealBlock, skip_ids: Set[int]):
+    slots_to_remove = set()
+    for i, op in enumerate(cur_block.ops[:-1]):
+        if type(op) != TealOp or op.op != Op.store:
+            continue
+
+        # do not optimize away reserved and global slots
+        if op.getSlots()[0].id in skip_ids:
+            continue
+
+        next_op = cur_block.ops[i + 1]
+        if type(next_op) != TealOp or next_op.op != Op.load:
+            continue
+
+        if op.args[0] != next_op.args[0]:
+            continue
+
+        if not _has_load_dependencies(cur_block, start, cast(int, op.args[0]), i + 1):
+            slots_to_remove.add(cast(int, op.args[0]))
+
+    _remove_extraneous_slot_access(start, slots_to_remove)
+
+
+def apply_global_optimizations(start: TealBlock, options: OptimizeOptions) -> TealBlock:
+    # limit number of iterations to length of teal program to avoid potential
+    # infinite loops.
+    for block in TealBlock.Iterate(start):
+        prev_ops = block.ops.copy()
+        for _ in range(len(block.ops)):
+            if options.scratch_slots:
+                _apply_slot_to_stack(block, start, options.skip_ids)
+
+            if prev_ops == block.ops:
```

Comment:
If this branch isn't true, I think you'll want to do `prev_ops = block.ops.copy()` again, otherwise you're not measuring the difference between iterations, you're measuring the difference from the beginning.